runtime/src/bank.rs

```diff
@@ -5806,6 +5810,61 @@ impl Bank {
             });
     }
 
+    /// After deserialize, populate skipped rewrites with accounts that would normally
+    /// have had their data rewritten in this slot due to rent collection (but didn't).
+    ///
+    /// This is required when starting up from a snapshot to verify the bank hash.
+    fn rebuild_skipped_rewrites(&mut self) {
+        // It is only legel to rebuild skipped rewrites when starting from a snapshot,
+        // which means the current `Self::skipped_rewrites` must be empty.
+        assert!(self.skipped_rewrites.get_mut().unwrap().is_empty());
+
+        let (skipped_rewrites, measure_skipped_rewrites) =
+            measure!(self.calculate_skipped_rewrites());
+        info!(
+            "Rebuilding skipped rewrites of {} accounts{measure_skipped_rewrites}",
+            skipped_rewrites.len()
+        );
+        *self.skipped_rewrites.get_mut().unwrap() = skipped_rewrites;
+    }
+
+    /// Calculates (and returns) skipped rewrites for this bank
+    ///
+    /// Refer to `rebuild_skipped_rewrites()` for more documentation.
+    /// This implementaion is purposely separate to facilitate testing.
+    ///
+    /// The key observation is that accounts in Bank::skipped_rewrites are only used IFF the
+    /// specific account is *not* already in the accounts delta hash.  If an account is not in
+    /// the accounts delta hash, then it means the account was not modified.  Since (basically)
+    /// all accounts are rent exempt, this means (basically) all accounts are unmodified by rent
+    /// collection.  So we just need to load the accounts that would've been checked for rent
+    /// collection, hash them, and add them to Bank::skipped_rewrites.
+    ///
+    /// As of this writing, there are ~350 million acounts on mainnet-beta.
+    /// Rent collection almost always collects a single slot at a time.
+    /// So 1 slot of 432,000, of 350 million accounts, is ~800 accounts per slot.
+    /// Since we haven't started processing anything yet, it should be fast enough to simply
+    /// load the accounts directly.
+    /// Empirically, this takes about 3-4 milliseconds.
+    fn calculate_skipped_rewrites(&self) -> HashMap<Pubkey, AccountHash> {
+        // The returned skipped rewrites may include accounts that were actually *not* skipped!
+        // (This is safe, as per the fn's documentation above.)
+        HashMap::from_iter(
+            self.rent_collection_partitions()
+                .into_iter()
+                .map(accounts_partition::pubkey_range_from_partition)
+                .flat_map(|pubkey_range| {
+                    self.rc
+                        .accounts
+                        .load_to_collect_rent_eagerly(&self.ancestors, pubkey_range)
```

Comment:
so this impl is duplicated code with rent collection, right? I think I'm ok with this. Just making sure we are comfortable with the potentially subtle differences over time since the code is copied. Note that this whole code path is only run at startup. Worst case is it causes nodes to fail at startup. Since we restart a small amount of validators at a time, if there was an inconsistency here we would know when each individual validator restarts. So, there is no risk of bringing down the cluster at steady state.

Reply:
Is this work when there are accounts in which rents are actually collected? 
It seems that in jeff's original code, those accounts are exclude from the skipped_rewrites?

Reply:
Yeah. I think so too. Thanks for confirming.